accepted/0000-package-aliases.md

```diff
@@ -0,0 +1,81 @@
+aliasing. While it can be somewhat inconvenient, it's still _possible_, and it
+doesn't seem to have stopped the ecosystem from growing as fast as it did. The
+simplicity of the npm model has been pretty good to us this whole time.
+
+Enter 2018, and we've started working on `npm asset`. After many conversations
+about it (with @isaacs nonetheless!) the CLI team finally decided to go ahead
+with this feature. Having the Yarn implementation around was a good reference
+when deciding _how_ to actually do this and totally helped make a case for it!
+
+## Detailed Explanation
+
+This will occur on two levels: one, a new spec type, `alias` will be added, to
+represent the `npm:<pkg>` part. The bigger change, though, will be allowing npm
+to install _any_ package under a different name, regardless of package type, and
+use the specific name provided when it adds it to `node_modules`. That is, `npm
+install foo@github:zkat/bar` will always install into a directory called `foo`,
+as will `npm install foo@npm:bar` and `npm i foo@../bar`.
+
+Furthermore, package IDs (used when comparing packages), will take aliases into
+account. For a package to be considered equivalent by npm, it must match BOTH
+the package name and version AS WELL AS the physical directory name it was
+installed under. This will take care of blocking aliases from injecting
+transitive dependencies.
+
+## Rationale and Alternatives
+
+Pretty much since the dawn of npm's time, there's been the question of "how do I
+install a package under this separate name?" -- be it for using git forks as
+replacements, or forked npm packages as replacements. This question is older
+than npm scopes themselves, so we've had some developments since the choice was
+originally made.
+
+There is no good, easy way to do this. You can, of course, craft special
+packages that re-export everything under a different name and install those, but
+that seems less than ideal in a lot of situations.
+
+The main use-cases for something like this seem to be:
+
+1. installing multiple versions of a same-name package side-by-side (`npm i jquery3@npm:jquery@3 jquery2@npm:jquery@2` -> installs `node_modules/jquery(2|3)`
+2. installing a registry-published fork under the name of the original package (`npm i npm@npm:@zkat/npm`)
+3. shorter, more convenient import names for packages that are otherwise longer than desired (`npm i npa@npm:npm-package-arg`)
+
+This RFC should solve all three of those problems.
+
+There are other, related problems that this RFC will NOT address:
+
+1. replacing packages in transitive dependencies that have different names (`npm i underscore@npm:lodash` will only make it so `require('underscore')` _in your own project_ loads `lodash` instead. Your dependencies will receive their own (nested) version of the actual `underscore` package). This may be covered in the future by an explicit resolutions mechanism, and may be facilitated by this RFC, but is still not part of it.
```

Comment:
Flattening of deps in npm 3+ seems like it will mean that this boundary isn’t in fact strictly or reliably adhered to. If a transitive dep depends on underscore at a version matching the aliased underscore, wouldn’t it use that one?

Reply:
no. That's exactly what this paragraph is saying. Dependencies are identified by _both_ name and spec version. `underscore@npm:lodash@1.0.0` and `underscore@1.0.0` are _entirely different_, and _incompatible_ dependencies, so they won't be flattened.

Reply:
ah ok gotcha, so the aliasing transparently applies to the filesystem folder name, but not to the "ideal dep graph" logic?

Reply:
both.